src/Components/Web.JS/src/Platform/Circuits/CircuitStartOptions.ts

```diff
@@ -35,7 +35,7 @@ export function resolveOptions(userOptions?: Partial<CircuitStartOptions>): Circ
 
 export interface ReconnectionOptions {
   maxRetries: number;
-  retryIntervalMilliseconds: number;
+  retryIntervalMilliseconds: number | ((attempt: number) => number);
```

Comment:
> This would allow apps that have an out-of-band way to check if a reconnect is likely to succeed to trigger a reconnect attempt immediately as soon as they think the server is ready without setting super short intervals.

That's an interesting idea. I wonder how common it would be in practice for customers to implement that. If we think it's not going to be common, we could suggest that customers implement their own `ReconnectionHandler` and use that instead of the default one:

```js
Blazor.start({
  reconnectionHandler: {
    onConnectionDown(options, error) { /* Custom fancy logic */ }
    onConnectionUp() { /* Stop custom fancy logic */ }
  },
});
```

That way we don't have to account for every possible use case in the `DefaultReconnectHandler` and just support the most common cases.

If we did add an API that allowed you to return a promise to decide when to attempt a reconnect, I would propose we not overload the `retryIntervalMilliseconds` option with that, simply because the relationship between the property name and its value stops being clear at that point.

Accepting an array for `retryIntervalMilliseconds` is also nice, but IMO opting to keep the number of types of acceptable values to a minimum is fine. If we go with your other suggestion where returning `null` or `undefined` from the `retryIntervalMilliseconds` function means "stop retrying", then it just becomes:

```js
retryIntervalMilliseconds: Array.prototype.at.bind([, 0, 1000, 1000, 1000, 2000, 10000, 30000])
```

What do you think?

Reply:
Fair point about the naming of `retryIntervalMilliseconds` making it particularly weird to allow returning promises from the function. However, I'm not sure that suggesting that customers implement their own `ReconnectionHandler` is an alternative to a promise-returning retry function. That seems useful for configuring the UI when the circuit goes up and down, but it doesn't allow me to trigger a reconnect attempt immediately when some out-of-band event triggers.

Imagine my modal reconnect UI has a button to retry the reconnection immediately. This could be particularly useful if enough attempts have already failed that the backoff logic has set the next interval all the way up to 30 seconds. If a person could click a button to retry immediately when they know their internet is back, that could be very useful. Maybe it would even be a good idea to provide a button like this by default, but I don't think it possible to do this with a custom `ReconnectionHandler`, because while you'll get notified of `onConnectionDown` and `onConnectionUp` events, I don't see any API to `attemptReconnectNow()`.

I still like returning `null` or `undefined` from the function meaning "stop retrying", but unless the function is actually overwritten to something custom, we should still respect `maxRetries`. That means default function would probably have to be a little more complicated than `Array.at`, because it would need to read the `maxRetries` configuration so it can return `null` at the appropriate time.

Reply:
>  but I don't think it possible to do this with a custom ReconnectionHandler, because while you'll get notified of onConnectionDown and onConnectionUp events, I don't see any API to attemptReconnectNow().

We have the `Blazor.reconnect` API that you can call at any time to attempt a reconnection, so it actually is possible. The default reconnect logic is defined as a `ReconnectionHandler` and anything the default handler does is something users should be able to do as well, if they're willing to write enough code.

> That means default function would probably have to be a little more complicated than `Array.at`, because it would need to read the `maxRetries` configuration so it can return null at the appropriate time.

I think it's okay if the default retry function is a bit longer. I already changed it to not use an array anyway. I was more saying that the customer could write their retry interval calculation like that.